Sources/Prometheus/PrometheusCollectorRegistry.swift

```diff
@@ -145,26 +152,66 @@ public final class PrometheusCollectorRegistry: Sendable {
             guard let entry = store[name] else {
                 // First time a Counter is registered with this name.
                 let counter = Counter(name: name, labels: labels)
-                let counterWithHelp = MetricWithHelp(metric: counter, help: help)
-                let counterGroup = MetricGroup(
-                    metricsByLabelSets: [key: counterWithHelp]
+                let counterFamily = MetricFamily(
+                    metricsByLabelSets: labels.isEmpty ? [:] : [key: counter],
+                    help: help,
+                    metricUnlabeled: labels.isEmpty ? counter : nil
                 )
-                store[name] = .counter(counterGroup)
+                store[name] = .counter(counterFamily)
                 return counter
             }
+
             switch entry {
-            case .counter(var existingCounterGroup):
-                if let existingCounterWithHelp = existingCounterGroup.metricsByLabelSets[key] {
-                    return existingCounterWithHelp.metric
+            case .counter(var existingCounterFamily):
+                // Check mutual exclusivity constraints. While Prometheus wouldn't break with sharing a
+                // metric name for labeled and unlabeled variants, the client enforces that each unique
+                // metric name must either be labeled (allowing multiple label sets) or unlabeled
+                // in order to support correct Prometheus aggregations.
+                if existingCounterFamily.metricUnlabeled != nil && !labels.isEmpty {
```

Comment:
Can't we do this in the constructor of `MetricFamily`? otherwise there's still ways to construct an illegal Family by accident

Reply:
I'd probably make all propertires `let` in the family so we can't violate the rule by external mutation, and then make some method called amend or add that performs these checks inside there. This way we don't have a chance to get the type into illegal state